### PR TITLE
Hieu Nguyen - Bai tap 05

### PR DESCRIPTION
**Pull request by at-hieunguyen3**
-Xong bai 5.1, 5.2, 5.4

### CODE REVIEW

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/BadgeButton.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+
+    convenience init(frame: CGRect, pointNumber: Int, name: String, colorBtn: UIColor, position: PositionBadge) {
+        self.init(frame: frame)
+        //heightButton = frame.size.height
+//        widthButton = frame.size.width
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/BadgeButton.swift:53]: Remove if no need

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/BadgeButton.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        self.layer.cornerRadius = 15
+        self.setTitle(name, for: .normal)
+        self.setTitleColor(.white, for: .normal)
+        self.backgroundColor = colorBtn
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/BadgeButton.swift:57]: No need self

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/BadgeButton.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        clipsToBounds = false
+        setupBadgeView(position: position, number: pointNumber)
+    }
+
+    func setupBadgeView(position: PositionBadge, number: Int) {
+        if number == 0 { return }
+        let sizeButton: CGSize = self.frame.size
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/BadgeButton.swift:64]: No need self

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+class MyAvatar: UIView {
+    
+    weak var delegate: MyAvatarDelegate?
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+    
+    convenience init(frame: CGRect,index: Int) {
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift:19]: Should init a avatar view with index and name, after that use this name to show

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        profileView.addSubview(nameLabel)
+        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(printInfo(_:)))
+        tapGesture.name = String(index)
+        profileView.addGestureRecognizer(tapGesture)
+        profileView.isUserInteractionEnabled = true
+        addSubview(profileView)
+    }
+
+    @objc private func printInfo(_ gesture: UITapGestureRecognizer) {
+        delegate?.setData(self, gesture.name!)
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift:53]: Should not use force unwrap

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        self.init(frame: frame)
+        setupSlider(value: Int(numberValue))
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func setupSlider(value: Int) {
+        print(bounds)
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:33]: Remove print

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        backgroundSlider.image = UIImage(named: "backgroundSlider")
+        backgroundSlider.contentMode = .top
+        backgroundSlider.frame = bounds
+        self.addSubview(backgroundSlider)
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:37]: No need self

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+        valueSlider.image = UIImage(named: "valueSlider")
+        valueSlider.frame = bounds
+        valueSlider.contentMode = .top
+        valueSlider.clipsToBounds = true
+        valueSlider.frame.size.height = bounds.size.height / 2
+        self.addSubview(valueSlider)
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:44]: No need self

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        
+        thumbSlider.frame.size = CGSize(width: 150, height: 50)
+        thumbSlider.center = CGPoint(x: bounds.size.width / 2, y: bounds.size.height / 2)
+        thumbSlider.backgroundColor = #colorLiteral(red: 0.9529411793, green: 0.6862745285, blue: 0.1333333403, alpha: 1)
+        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(handleSlider))
+        thumbSlider.isUserInteractionEnabled = true
+        thumbSlider.addGestureRecognizer(panGesture)
+        self.addSubview(thumbSlider)
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:52]: No need self

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        
+        numberLabel.text = "\(value)"
+        numberLabel.textColor = .white
+        numberLabel.frame = thumbSlider.bounds
+        numberLabel.textAlignment = .center
+        thumbSlider.addSubview(numberLabel)
+        
+        print(value)
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:60]: Remove

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    @objc func handleSlider(_ tapGesture: UIPanGestureRecognizer) {
+        let sliderY = self.bounds.origin.y
+        let sliderHeight = self.bounds.height
+        let heightValue = tapGesture.view!.center.y
+        print("\(sliderY) \(sliderHeight) \(heightValue)")
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:66]: Remove

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        numberLabel.textColor = .white
+        numberLabel.frame = thumbSlider.bounds
+        numberLabel.textAlignment = .center
+        thumbSlider.addSubview(numberLabel)
+        
+        print(value)
+    }
+    @objc func handleSlider(_ tapGesture: UIPanGestureRecognizer) {
+        let sliderY = self.bounds.origin.y
+        let sliderHeight = self.bounds.height
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:64]: No need self

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        let heightValue = tapGesture.view!.center.y
```
Comment (at-bienle) [inline 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:65]: Should not use force unwrap

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift (hunk 5 above)
Comment (at-hieunguyen3) [reply 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift:53]: I edited by using safe force unwrap

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift (hunk 13 above)
Comment (at-hieunguyen3) [reply 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/SliderView.swift:65]: I edited by using safe force unwrap

File: 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift (hunk 4 above)
Comment (at-hieunguyen3) [reply 05_CustomView/HieuNguyen/BaiTap05/BaiTap05/View/PersonView.swift:19]: Because "Name + index" is required,so i think each index is enough

**Review by at-bienle — APPROVED** [review]
LGTM!

**Pull request merged**